shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs

```diff
@@ -1085,23 +1104,123 @@ updateNonMyopic nm rPot newLikelihoods =
 
--- | Create a reward update
-createRUpd ::
-  forall era.
+-- =========================================
+-- Useful data to pass around intermediate results
+
+-- | To pulse the reward update, we need a snap shot of the EpochState particular to this computation
+data RewardSnapShot crypto = RewardSnapShot
+  { rewSnapshots :: !(SnapShots crypto),
+    rewa0 :: Rational,
+    rewnOpt :: Natural,
+    rewprotocolVersion :: ProtVer,
+    rewNonMyopic :: !(NonMyopic crypto),
+    rewDeltaR1 :: !Coin, -- deltaR1
+    rewR :: !Coin, -- r
+    rewDeltaT1 :: !Coin, -- deltaT1
+    rewTotalStake :: !Coin, -- totalStake
+    rewRPot :: !Coin -- rPot
+  }
+  deriving (Show, Eq, Generic)
+
+-- instance HasField "_d" (RewardSnapShot crypto) UnitInterval where
+-- instance HasField "_tau" (RewardSnapShot crypto) UnitInterval where
+-- instance HasField "_rho" (RewardSnapShot crypto) UnitInterval where
+instance HasField "_a0" (RewardSnapShot crypto) Rational where
+  getField x = rewa0 x
+
+instance HasField "_nOpt" (RewardSnapShot crypto) Natural where
+  getField x = rewnOpt x
+
+instance HasField "_protocolVersion" (RewardSnapShot crypto) ProtVer where
+  getField x = rewprotocolVersion x
+
+instance NoThunks (RewardSnapShot crypto)
+
+instance NFData (RewardSnapShot crypto)
+
+instance CC.Crypto crypto => ToCBOR (RewardSnapShot crypto) where
+  toCBOR (RewardSnapShot ss a0 nopt ver nm dr1 r dt1 tot pot) =
+    encode
+      ( Rec RewardSnapShot !> To ss !> To a0 !> To nopt !> To ver !> To nm !> To dr1
+          !> To r
+          !> To dt1
+          !> To tot
+          !> To pot
+      )
+
+instance CC.Crypto crypto => FromCBOR (RewardSnapShot crypto) where
+  fromCBOR = decode (RecD RewardSnapShot <! From <! From <! From <! From <! From <! From <! From <! From <! From <! From)
+
+-- | State used in the STS rules
+data PulsingRewUpdate m crypto
+  = Pulsing !(RewardSnapShot crypto) !(RewardPulser m crypto) -- Pulsing work to do
+  | Complete !(RewardUpdate crypto) -- Pulsing work completed
+  deriving (Eq, Show, Generic, NoThunks)
+
+instance NFData (PulsingRewUpdate m crypto)
+
+instance (Typeable m, CC.Crypto crypto) => ToCBOR (PulsingRewUpdate m crypto) where
+  toCBOR (Pulsing s p) = encode (Sum Pulsing 0 !> To s !> To p)
+  toCBOR (Complete r) = encode (Sum Complete 1 !> To r)
+
+instance (Monad m, Typeable m, CC.Crypto crypto) => FromCBOR (PulsingRewUpdate m crypto) where
+  fromCBOR = decode (Summands "PulsingRewUpdate" decPS)
+    where
+      decPS 0 = (SumD Pulsing <! From <! From)
+      decPS 1 = (SumD Complete <! From)
+      decPS n = Invalid n
+
+-- ====================================================================
+-- Some generic lifting functions to lift one provenance computation
+-- into another, and one to lift a Pulser from one provenance type to
+-- another.  Then a specialisation on the Provenance types we use here.
+
+-- | Lift a Pulser in the ProvM monad, from one type of provenance (s1) to another (s2)
+pulseProvM ::
```

Comment:
Avoid putting this stuff in `LedgerState`

Reply:
Move to `RewardProvenance`

Reply:
done